src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs

```diff
@@ -3,140 +3,106 @@
+                    var ambiguousMethods = string.Join(", ", ambiguousMatches.Select(m => m.Method));
+                    throw new InvalidOperationException(Resources.FormatAmbiguousHandler(Environment.NewLine, ambiguousMethods));
                 }
 
-                if (best != null)
+                if (bestMatch != null)
                 {
-                    return best.Value.Handler;
+                    return bestMatch;
                 }
             }
 
             return null;
         }
 
-        // Bad prototype substring implementation :)
-        private struct HandlerMethodAndMetadata
+        private static List<HandlerMethodDescriptor> SelectHandlers(PageContext context)
         {
-            public static HandlerMethodAndMetadata Create(HandlerMethodDescriptor handler)
-            {
-                var name = handler.Method.Name;
+            var handlers = context.ActionDescriptor.HandlerMethods;
+            List<HandlerMethodDescriptor> handlersToConsider = null;
 
-                string httpMethod;
-                if (name.StartsWith("OnGet", StringComparison.Ordinal))
+            var formAction = Convert.ToString(context.RouteData.Values[FormAction]);
+            for (var i = 0; i < handlers.Count; i++)
+            {
+                var handler = handlers[i];
+                if (handler.HttpMethod != null &&
+                    !string.Equals(handler.HttpMethod, context.HttpContext.Request.Method, StringComparison.OrdinalIgnoreCase))
                 {
-                    httpMethod = "GET";
+                    continue;
                 }
-                else if (name.StartsWith("OnPost", StringComparison.Ordinal))
+                else if (handler.FormAction.HasValue &&
+                    !handler.FormAction.Equals(formAction, StringComparison.OrdinalIgnoreCase))
                 {
-                    httpMethod = "POST";
+                    continue;
                 }
-                else
+
+                if (handlersToConsider == null)
                 {
-                    httpMethod = null;
+                    handlersToConsider = new List<HandlerMethodDescriptor>();
                 }
 
-                var formactionStart = httpMethod?.Length + 2 ?? 0;
-                var formactionLength = name.EndsWith("Async", StringComparison.Ordinal)
-                    ? name.Length - formactionStart - "Async".Length
-                    : name.Length - formactionStart;
+                handlersToConsider.Add(handler);
+            }
 
-                var formaction = formactionLength == 0 ? null : name.Substring(formactionStart, formactionLength);
+            return handlersToConsider;
+        }
 
-                return new HandlerMethodAndMetadata(handler, httpMethod, formaction);
+        private static int GetScore(HandlerMethodDescriptor descriptor)
```

Comment:
Could/should we include a comment either here or in the calling function giving a little description about what exactly these scores mean and how they're used? It feels a tad opaque to me.